### ISSUE TITLE
Investigate why certain Browse API requests are delayed

### ISSUE THREAD

**Issue by kumar303** [mozilla/addons-code-manager#1320]
### Describe the problem and steps to reproduce it:

* Open the Network panel
* Load any browse page, such as https://code.addons-dev.allizom.org/en-US/browse/616959/versions/1688394/?path=popup%2Fquicknote.js

### What happened?

The API calls to fetch the version file and the version are: 
* run sequentially
* only begin after the user profile API call

### What did you expect to happen?

Since the version / version file API calls are crucial to display code on the page, they should be given priority and should run in parallel.

### Anything else we should know?

<img width="1986" alt="Screenshot 2019-12-06 10 25 54" src="https://user-images.githubusercontent.com/55398/70338792-2e646c80-1813-11ea-833c-062a5b7c51c9.png">

Also see https://github.com/mozilla/addons-code-manager/issues/1300

**Comment by kumar303** [comment]
In the past I've advised against the [prensentation / container pattern](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0) because it discourages autonomous, encapsulated components but it might be a good strategy for prioritizing crucial API requests. That's one thing I hadn't considered.

**Comment by willdurand** [comment]
On the screenshot above, we also see that the version API is called twice to retrieve the same file. That does sound like a good idea.

**Comment by willdurand** [comment]
@bobsilverberg I took a look at what I described in my last comment (above). Reading the code, there is a possible race condition in the `Browse` component when we load a version and a file.

In `loadData()`, here are the steps:

1. `_fetchVersion` with a version ID and a file
2. the `_fetchVersion` thunk calls the API
3. the thunk loads the version info
4. the thunk loads the version file

Now, if we "zoom" between (3) and (4):

3. the thunk loads the version info
  - the redux state is updated, therefore it triggers a re-render
  - the `loadData()` is called
  - in `loadData()`, `_fetchVersion()` is skipped because a `version` exists
  - in `loadData()`, there is no `file` loaded yet but the `selectedPath` has been set already and we are not currently loading a file, so `_fetchVersionFile()` is called 💣
4. the thunk loads the version file

This naive patch fixes the issue (i.e. only one call to the version API is performed) and the test suite passes:

```diff
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index acf5c2e..468e2cb 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -815,13 +815,13 @@ export const fetchVersion = ({
       dispatch(actions.abortFetchVersion({ versionId }));
       dispatch(errorsActions.addError({ error: response.error }));
     } else {
-      dispatch(actions.loadVersionInfo({ version: response }));
       dispatch(
         actions.loadVersionFile({
           version: response,
           path: response.file.selected_file,
         }),
       );
+      dispatch(actions.loadVersionInfo({ version: response }));
     }
   };
 };
```

The idea behind this patch is that:

1. `_fetchVersion()` can be called multiple times, it is "protected" to make sure we only run it once at a time (thanks to `selectVersionIsLoading()`)
2. `loadVersionFile()` does not need any "version info"
3. calling `loadVersionFile()` first will also trigger a re-render but since there is no version (info) loaded yet, that will early return after calling `_fetchVersion()` and this thunk will early return as well
4. loading the version info after will trigger a re-render and here we have everything: version info and file

**Comment by bobsilverberg** [comment]
Thanks for looking into this @willdurand. I hadn't seen it because I was focussing on `mvp` issues at the top of the project queue, but I think performance *is* an `mvp` issue, so I've added that label and I think we *should* treat this as a high priority.

At first glance your suggestion above makes sense, but I want to take a closer look, as I'm wondering if we need to introduce a second loading flag, similar to what's used with `selectVersionIsLoading()`, or maybe change the logic for "VersionIsLoading" to deal with `VersionFile` as well. I'll take a closer look later today.

**Comment by bobsilverberg** [comment]
Ok, so we do already have `versionFilesLoading` in state, but it must be set to `false` at the point at which we encounter `_fetchVersionFile()` in `loadData()`, otherwise that call to `_fetchVersionFile()` wouldn't be made. Does this indicate a logic error somewhere? Perhaps `versionFilesLoading` needs to be set to `true` when `actions.beginFetchVersion` executes, and maybe that would also resolve the issue?

Your patch is good, but I just want to make sure that this double API call isn't a result of a logic error on our part with our code that is designed to handle this stuff.

@willdurand What do you think?

**Comment by willdurand** [comment]
I believe it is a race condition so not a logic error _per se_.

Given that we are loading a file along with the version info, we might need to set the `versionFilesLoading` to `true` in `fetchVersion()`, unless we use `fetchVersion()` without loading a file.

Also, in `fetchVersion()` we early return if the boolean flag is `true` but we are not doing this in `fetchVersionFile()`.

**Comment by bobsilverberg** [comment]
> I believe it is a race condition so not a logic error _per se_.
> 
> Given that we are loading a file along with the version info, we might need to set the `versionFilesLoading` to `true` in `fetchVersion()`, unless we use `fetchVersion()` without loading a file.
> 
> Also, in `fetchVersion()` we early return if the boolean flag is `true` but we are not doing this in `fetchVersionFile()`.

The point I was trying to make was that we should make sure we are using `versionFilesLoading` correctly and fully, and your comment directly above suggests that that is not the case.

If we do that and still have a race condition then we can address that, perhaps with your suggested patch, but I do think we should make sure we're using the features we put in place correctly first.

**Comment by willdurand** [comment]
Right. As I wrote in the OP, it was a naive patch. I wanted to start a discussion and what you said makes perfectly sense. Let's start by protecting `fetchVersionFile()` first.

**Comment by bobsilverberg** [comment]
> Right. As I wrote in the OP, it was a naive patch. I wanted to start a discussion and what you said makes perfectly sense. Let's start by protecting `fetchVersionFile()` first.

Agreed! Do you want to open a separate issue about that?

**Comment by willdurand** [comment]
> Do you want to open a separate issue about that?

Yes

**Comment by willdurand** [comment]
> Agreed! Do you want to open a separate issue about that?

I filed https://github.com/mozilla/addons-code-manager/issues/1338
